### PR TITLE
Add triplet loss

### PR DESCRIPTION
**Pull request by MARD1NO**
增加 MarginRankingLoss 和 Triplet Loss（因为后者需要用到marginloss所以一并添加）

任务:
 - [x] : 增加 MarginLoss
 - [x] : 增加 MarginLoss前向测试
 - [x] : 增加 MarginLoss梯度验证
 - [x] : 增加 TripletLoss
 - [x] : 增加 TripletLoss前向测试
 - [x] : 增加 TripletLoss梯度验证
## Margin Loss
公式为
image

分别对应reduction = "none", "sum", "mean" 三种情况

## Triplet Loss
对齐 `torch.TripletMarginLoss`，
源代码可参考：
[triplet loss](https://github.com/pytorch/pytorch/blob/df0ae244a999a365ac13695cf4682ea30490b157/aten/src/ATen/native/Loss.cpp#L62)

[distance](https://github.com/pytorch/pytorch/blob/df0ae244a999a365ac13695cf4682ea30490b157/aten/src/ATen/native/Distance.cpp#L15)

原公式为
图片

我们最常见的p2范数，公式为
图片

但是对于其他范数，我们通过 `flow.math.pow` 去组合得到，其中的一个问题是特殊的如 `pow(-1, 1/3)` 这是个不合法的，参照torch的源代码，**它在前面做了个取绝对值的操作**，保证输入的都是正数，进而**求到的范数是合法的**。

可参考 other那一栏
图片

# 求导过程
我们先单看 d(a-p)这项的求导，导数为grad1
image
同理 d(a-n)的求导也是这样，导数为grad2

我们梯度检查针对a，所以是 grad1- grad2
由于范数里面涉及到了绝对值，所以最后外面需要乘一个绝对值操作得到的导数 abs_grad，具体可以参考单测代码

# 注意事项
暂时没有支持 swap = True

原因是我们的 `oneflow.math.minimum` 缺乏反向传播，所以设置了个assert强行指定为False

### CODE REVIEW

File: oneflow/python/ops/nn_ops.py [hunk 1]
```diff
@@ -2646,3 +2646,264 @@ def leaky_relu_Job(x: tp.Numpy.Placeholder((5, ),)
...
+            x ([type]): The input Blob. 
+            p ([type], optional): The norm degree. Defaults to 2..
+        """
+        # In order to avoid the `nan` case.
+        _abs_val = flow.math.abs(x, name=name + "_abs")
+
+        if p == 2.0:
+            # Use Square to compute the p2-norm
+            _norm = flow.math.square(_abs_val, name=name + "_square")
+            _norm = flow.math.reduce_sum(_norm, axis=-1, name=name + "_sum")
```
Comment (Ldpe2G) [inline oneflow/python/ops/nn_ops.py:2871]: triplet loss 默认输入是只有两维的是吧，我看了下 pytorch 的代码，貌似reduce 的维度默认是在第一维，如果是两维的就是一样的。

Comment (MARD1NO) [reply oneflow/python/ops/nn_ops.py:2871]: 拿代码试验了下还真是，这个我改下

Comment (Ldpe2G) [reply oneflow/python/ops/nn_ops.py:2871]: 先看到 `pairwise_distance`
```python3
Tensor pairwise_distance(const Tensor& x1, const Tensor& x2, double p, double eps, bool keepdim) {
  return at::norm(x1 - x2 + eps, p, 1, keepdim);
}
```
https://github.com/pytorch/pytorch/blob/25ba802ce4cbdeaebcad4a03cec8502f0de9b7b3/aten%2Fsrc%2FATen%2Fnative%2FDistance.cpp#L15

再跟进去看 `norm`，可以看到dim传的是1

```python3
static Tensor norm(const Tensor& self, optional<Scalar> p, IntArrayRef dim, bool keepdim,
            optional<ScalarType> opt_dtype) {
  if (self.is_sparse()) {
    // Sparse tensors need a different implementation because their values
    // are accessed with a different API than strided tensors
    return at::native_norm(self, p, dim, keepdim, opt_dtype);
  } else {
    Tensor result;
    return at::native::norm_out(result, self, p, dim, keepdim, opt_dtype);
  }
}
```
https://github.com/pytorch/pytorch/blob/ab0a04dc9c8b84d4a03412f1c21a6c4a2cefd36c/aten%2Fsrc%2FATen%2Fnative%2FReduceOps.cpp#L513

File: oneflow/python/test/ops/test_TripletMarginLoss.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+
+def _compare_triplet_margin_loss_with_np(
+    anchor_shape,
+    pos_shape,
+    neg_shape,
+    margin,
+    p,
+    swap,
+    device_type,
+    machine_ids,
```
Comment (Ldpe2G) [inline oneflow/python/test/ops/test_TripletMarginLoss.py:34]: 同样没用上

File: oneflow/python/test/ops/test_TripletMarginLoss.py [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+
+    flow.clear_default_session()
+    if device_type == "cpu":
+        flow.config.cpu_device_num(device_counts)
+    else:
+        flow.config.gpu_device_num(device_counts)
+
+    func_config = flow.FunctionConfig()
+
+    def np_triplet_margin_loss(np_anchor, np_pos, np_neg, np_margin, np_p, swap):
```
Comment (Ldpe2G) [inline oneflow/python/test/ops/test_TripletMarginLoss.py:51]: np_margin 没用上

File: oneflow/python/test/ops/test_TripletMarginLoss.py [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+        np_d_1_norm = np.power(np.abs((np_anchor - np_pos + 1e-6)), np_p)
+        np_d_2_norm = np.power(np.abs((np_anchor - np_neg + 1e-6)), np_p)
+
+        np_d_1 = np.power(np.sum(np_d_1_norm, axis=-1), 1.0 / p)
```
Comment (Ldpe2G) [inline oneflow/python/test/ops/test_TripletMarginLoss.py:55]: 应该是 np_p ?

File: oneflow/python/test/ops/test_TripletMarginLoss.py [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+    flow.clear_default_session()
+    if device_type == "cpu":
+        flow.config.cpu_device_num(device_counts)
+    else:
+        flow.config.gpu_device_num(device_counts)
+
+    func_config = flow.FunctionConfig()
+
+    def np_triplet_margin_loss(np_anchor, np_pos, np_neg, np_margin, np_p, swap):
+        np_d_1_norm = np.power(np.abs((np_anchor - np_pos + 1e-6)), np_p)
```
Comment (Ldpe2G) [inline oneflow/python/test/ops/test_TripletMarginLoss.py:52]: eps 用参数传递吧，不要硬编码

File: oneflow/python/test/ops/test_MarginRankingLoss.py [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+
+    def np_margin_ranking_diff(np_out, np_target):
+        # Use numpy to compute diff
+        # Here we only test the input_1 gradient
+        # If loss > 0, the grad is: -target, else the grad is 0
+        elem_cnt = np_out.size
+        row, col = np_out.shape
+        np_diff = np.zeros(shape=np_out.shape)
+        # TODO: Optimize the backward logic
+        for i in range(row):
```
Comment (Ldpe2G) [inline oneflow/python/test/ops/test_MarginRankingLoss.py:71]: 这样写就局限是两维的输入了，建议看下能不能用上numpy的向量操作，或者直接flatten为一维处理

File: oneflow/python/test/ops/test_TripletMarginLoss.py [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+    def np_triplet_loss_diff(anchor, pos, neg, margin, p):
+        def _compute_distance(x1, x2, x3):
+            d_1_norm = np.power(np.abs((x1 - x2 + 1e-6)), p)
+            d_2_norm = np.power(np.abs((x1 - x3 + 1e-6)), p)
+            d_1 = np.power(np.sum(d_1_norm, axis=-1), 1.0 / p)
+            d_2 = np.power(np.sum(d_2_norm, axis=-1), 1.0 / p)
+
+            return d_1 - d_2 + margin
+
+        def _compute_per_diff(x1, x2, p, eps=1e-6):
```
Comment (Ldpe2G) [inline oneflow/python/test/ops/test_TripletMarginLoss.py:84]: 你上面写的给的梯度公式中间应该是没有求和符号的，不过实现上看起来是正确的，但是下面 allclose 的误差需要放宽到 1e-3，这个还要再检查下是不是哪里细节不对。

image

Comment (MARD1NO) [reply oneflow/python/test/ops/test_TripletMarginLoss.py:84]: 这个我没仔细一个个尝试，默认1e-5有时候会过不了。

File: oneflow/python/test/ops/test_MarginRankingLoss.py (hunk 6 above)
Comment (MARD1NO) [reply oneflow/python/test/ops/test_MarginRankingLoss.py]: 已修改

File: oneflow/python/test/ops/test_MarginRankingLoss.py [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+
+    @flow.global_function(
+        type="train", function_config=func_config,
+    )
+    def oneflow_marginloss(
+        of_input1: tp.Numpy.Placeholder(shape=input1.shape),
+        of_input2: tp.Numpy.Placeholder(shape=input2.shape),
+        of_target: tp.Numpy.Placeholder(shape=target.shape),
+    ) -> Dict[str, tp.Numpy]:
+        with flow.scope.placement(device_type, "0:0"):
```
Comment (Ldpe2G) [inline oneflow/python/test/ops/test_MarginRankingLoss.py:102]: 这里应该是不需要单独为varaible设置placement的

Comment (MARD1NO) [reply oneflow/python/test/ops/test_MarginRankingLoss.py:102]: 这里加 Placement的原因是，如果不单独将watch diff的变量放在一张卡上，跑单机多卡的时候，watch_diff会报错。
这里仅仅是为了单机多卡的watch_diff 运行正确

Comment (Ldpe2G) [reply oneflow/python/test/ops/test_MarginRankingLoss.py:102]: 这算是bug吗

Comment (Ldpe2G) [reply oneflow/python/test/ops/test_MarginRankingLoss.py:102]: 我知道了，foreign_watch_op 要求 parallel_num 要是1

File: oneflow/python/ops/nn_ops.py [hunk 9]
```diff
@@ -2946,3 +2948,251 @@ def mseloss_job(input: tp.Numpy.Placeholder(shape=(3, 3)),
...
+    assert reduction in [
+        "none",
+        "mean",
+        "sum",
+    ], "{} is not a valid value for reduction, The reduction must be the one of `none`, `mean`, `sum`. ".format(
+        reduction
+    )
+
+    if name is None:
+        name = "MarginRankingLoss_"
```
Comment (Ldpe2G) [inline oneflow/python/ops/nn_ops.py:3038]:
```python3
name = id_util.UniqueStr("MarginRankingLoss_")
```

File: oneflow/python/ops/nn_ops.py [hunk 10]
```diff
@@ -2946,3 +2948,251 @@ def mseloss_job(input: tp.Numpy.Placeholder(shape=(3, 3)),
...
+        name (Optional[str], optional): The name for the operation. Defaults to None.
+    
+    Returns:
+        remote_blob_util.BlobDef: The result Blob. 
+    """
+    assert (
+        input1.shape == input2.shape
+    ), "The shape of `input1`, `input2` must be the same. "
+
+    # TODO: Should we add a check of target?
```
Comment (Ldpe2G) [inline oneflow/python/ops/nn_ops.py:3027]: TODO(zheng zekang): Should we add a check of target?

File: oneflow/python/ops/nn_ops.py [hunk 11]
```diff
@@ -2946,3 +2948,251 @@ def mseloss_job(input: tp.Numpy.Placeholder(shape=(3, 3)),
...
+    ], "{} is not a valid value for reduction, The reduction must be the one of `none`, `mean`, `sum`. ".format(
+        reduction
+    )
+
+    assert (
+        swap == False
+    ), "For now we only support `swap=True`, OneFlow still have backward error in minimum"
+
+    if name is None:
+        name = "TripletMarginLoss_"
```
Comment (Ldpe2G) [inline oneflow/python/ops/nn_ops.py:3143]: name = id_util.UniqueStr("TripletMarginLoss_")

File: oneflow/python/ops/nn_ops.py [hunk 12]
```diff
@@ -2946,3 +2948,251 @@ def mseloss_job(input: tp.Numpy.Placeholder(shape=(3, 3)),
...
+            out = \sqrt[P]{\sum_{i=0}^{n}(abs(x)^P)} 
+        Args:
+            x ([type]): The input Blob. 
+            p ([type], optional): The norm degree. Defaults to 2..
+        """
+        # In order to avoid the `nan` case.
+        _abs_val = flow.math.abs(x, name=name + "_abs")
+
+        if p == 2.0:
+            # Use Square to compute the p2-norm
```
Comment (Ldpe2G) [inline oneflow/python/ops/nn_ops.py:3158]: # Use Square to compute the l2-norm

File: oneflow/python/ops/nn_ops.py [hunk 13]
```diff
@@ -2946,3 +2948,251 @@ def mseloss_job(input: tp.Numpy.Placeholder(shape=(3, 3)),
...
+
+    # Compute the distance
+
+    _distance_1 = _p_norm(anchor - positive + eps, p=p, name=name + "_distance_1")
+    _distance_2 = _p_norm(anchor - negative + eps, p=p, name=name + "_distance_2")
+
+    if swap:
+        _distance_swap = _p_norm(positive - negative + eps, p=p)
+        _distance_swap = flow.math.reduce_sum(_distance_swap, axis=1)
+        # TODO: minimum still not support backward
```
Comment (Ldpe2G) [inline oneflow/python/ops/nn_ops.py:3188]: TODO(zheng zekang): ...

**Review by Ldpe2G — APPROVED** [review]
